### PR TITLE
feat: handle NaN values in argmin

### PR DESCRIPTION
**Pull request by wjones127**
Currently, we panic if NaN is found in the array argmin or argmax operate on.

Example where we can hit this:

```python
import lance
import pyarrow as pa
import pyarrow.compute as pc
import shutil


dims = 375
nrows = 18836
# all zeros
inner_vector = pc.cast(pa.array([0] * (nrows * dims)), pa.float32())
vector = pa.FixedSizeListArray.from_arrays(inner_vector, dims)
table = pa.Table.from_arrays([vector], ["vector"])

shutil.rmtree("test", ignore_errors=True)
dataset = lance.write_dataset(table, "test")

dataset.create_index(
    "vector",
    index_type="IVF_PQ",
    metric="cosine",
    num_partitions=4,
    num_sub_vectors=125,
)
```

### CODE REVIEW

File: rust/src/arrow/kernels.rs [hunk 1]
```diff
@@ -34,15 +35,25 @@ pub fn argmax<T: ArrowNumericType>(array: &PrimitiveArray<T>) -> Option<u32>
...
+        if let Some(value) = value {
+            if let Some((_, max_value)) = max {
+                match value.partial_cmp(&max_value) {
+                    Some(Ordering::Greater) => max = Some((idx as u32, value)),
+                    Some(Ordering::Equal) => {}
+                    _ => {}
+                }
+            } else {
+                // Ignore NaN values
+                if let Some(Ordering::Equal) = value.partial_cmp(&value) {
```
[the remark this answers is not in the document]
Comment (wjones127) [reply rust/src/arrow/kernels.rs:49]: `value` is not `None`, it is `NaN`. This is the only way I could think of to generically figure out if the value is `NaN`.

File: rust/src/arrow/kernels.rs [hunk 2]
```diff
@@ -55,6 +60,11 @@ where
     array
         .iter()
         .enumerate()
+        .filter(|(_, value)| match value {
+            // Remove None and NaN values
```
Comment (eddyxu) [inline rust/src/arrow/kernels.rs:64]: Given that this is at critical path for queries, could you add one benchmark to how much overhead this adds?

Comment (eddyxu) [reply rust/src/arrow/kernels.rs:64]: Is the `panic` from line 71?

```rust
(Some(vx), Some(vy)) => vy.partial_cmp(vx).unwrap(),
```

If so, can we just take care the unwrap there?

Comment (wjones127) [reply rust/src/arrow/kernels.rs:64]: I couldn't figure out how to handle the unwrap within `max_by`, but I can give it another shot.

File: rust/src/arrow/kernels.rs [hunk 3]
```diff
@@ -40,9 +40,21 @@ where
...
+                if let Some(ordering) = vx.partial_cmp(vy) {
+                    ordering
+                } else {
+                    Ordering::Greater
+                }
+            }
+        })
+        .and_then(|(idx, val)| {
+            // If the value is NaN, return None
+            if let Some(val) = val {
```
Comment (eddyxu) [inline rust/src/arrow/kernels.rs:53]: Is it easier / faster to just run a `for loop` here?

Comment (wjones127) [reply rust/src/arrow/kernels.rs:53]: for loop over what?

Comment (eddyxu) [reply rust/src/arrow/kernels.rs]: Just a loop to run `argmin` together.

```
let mut min_idx = None;
let min_value = T::Max;
for v in array {
   if let Some(val) = v {
       if let Some(o) = vx.partial_cmp(&min_val) {
          ...
       }
   }
}
```

Comment (wjones127) [reply rust/src/arrow/kernels.rs]: That would be a lot simpler, but there isn't any `T::Max`. Can't find anything like that in the arrow crate so far.

Comment (eddyxu) [reply rust/src/arrow/kernels.rs]: We could probally either make a new trait `MaxValue`

and change the signature to

```
T: ArrowNumericType + MaxValue
```

Or just limit `argmin` to floats?

```
T:  ArrowNumricType + num_traits::float::Float
```

Comment (wjones127) [reply rust/src/arrow/kernels.rs]: Just did the first option. LMK what you think.

File: rust/src/arrow/kernels.rs [hunk 4]
```diff
@@ -27,40 +27,67 @@ use arrow_schema::DataType;
...
+                Self::MIN
+            }
+        }
+    };
+}
+
+impl_numeric_max_min!(f32);
+impl_numeric_max_min!(f64);
+impl_numeric_max_min!(i16);
+impl_numeric_max_min!(u32);
```
Comment (eddyxu) [inline rust/src/arrow/kernels.rs:51]: Looks good, lets just fill them with the rest of the numeric types?

Maybe we should also add support for two 16bit floats too.

File: rust/src/arrow/kernels.rs [hunk 5]
```diff
@@ -27,40 +27,67 @@ use arrow_schema::DataType;
 
 use crate::{Error, Result};
 
+pub trait NumericMaxMin {
```
Comment (eddyxu) [inline rust/src/arrow/kernels.rs:30]: Add some docs?

File: rust/src/arrow/kernels.rs (hunk 4 above)
Comment (wjones127) [reply rust/src/arrow/kernels.rs]: Just realized this exact trait already exists in num_traits as `Bounded`, so replaced with that.

**Review by eddyxu — APPROVED** [review]
Nice work!

Pending CI

**Pull request merged**